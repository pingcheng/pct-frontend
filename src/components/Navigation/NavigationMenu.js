import React, { Component } from "react";
import { NavLink } from "react-router-dom";

export class NavigationMenu extends Component {

	constructor(props) {
		super(props);

		this.state = {
			isMobileMenuVisible: false
		};
	}

	toggleMobileMenu = () => {
		this.setState({
			isMobileMenuVisible: !this.state.isMobileMenuVisible
		});
	};

	render() {
		return (
			<nav className="bg-gray-700">
				<div className="container-body">
					<div className="relative flex items-center justify-between h-16">
						<div className="absolute inset-y-0 left-0 flex items-center sm:hidden">
							<button onClick={this.toggleMobileMenu} className="inline-flex items-center justify-center p-2 rounded-md text-gray-400 hover:text-white hover:bg-gray-700 focus:outline-none focus:ring-2 focus:ring-inset focus:ring-white" aria-expanded="false">
								<span className="sr-only">Open main menu</span>

								<svg className="block h-6 w-6" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" aria-hidden="true">
									<path strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M4 6h16M4 12h16M4 18h16" />
								</svg>

								<svg className="hidden h-6 w-6" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" aria-hidden="true">
									<path strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M6 18L18 6M6 6l12 12" />
								</svg>
							</button>
						</div>

						<div className="flex-1 flex items-center justify-center sm:items-stretch sm:justify-start">

							<div className="flex-shrink-0 flex items-center">
								<div className="font-bold text-white" style={{
									borderTop: "3px solid white",
									borderBottom: "3px solid white"
								}}>Ping Cheng</div>
							</div>

							<div className="hidden sm:block sm:ml-6">
								<div className="flex space-x-4">
									<NavLink to="/" exact className="nav-item" activeClassName="active">Home</NavLink>
									<NavLink to="/posts" className="nav-item" activeClassName="active">Posts</NavLink>
									<NavLink to="/portfolio" className="nav-item" activeClassName="active">Portfolio</NavLink>
									<NavLink to="/about" className="nav-item" activeClassName="active">About me</NavLink>
								</div>
							</div>
						</div>
					</div>
				</div>

				<div className={`sm:hidden ${this.state.isMobileMenuVisible ? "block" : "hidden"}`}>
					<div className="px-2 pt-2 pb-3 space-y-1">
						<NavLink to="/" exact className="nav-item" activeClassName="active">Home</NavLink>
						<NavLink to="/posts" className="nav-item" activeClassName="active">Posts</NavLink>
						<NavLink to="/portfolio" className="nav-item" activeClassName="active">Portfolio</NavLink>
						<NavLink to="/about" className="nav-item" activeClassName="active">About me</NavLink>
					</div>
				</div>
			</nav>
		)
	}
}