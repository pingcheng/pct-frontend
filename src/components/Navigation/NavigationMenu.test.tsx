import { render, screen, waitFor } from "@testing-library/react";
import { NavigationMenu } from "./NavigationMenu";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

describe("test <NavigationMenu>", () => {
  it("should render correctly", function () {
    const { container } = render(
      <MemoryRouter>
        <NavigationMenu />
      </MemoryRouter>
    );
    expect(container.innerHTML).toMatchSnapshot();
  });

  it("should render mobile view toggle button", function () {
    render(
      <MemoryRouter>
        <NavigationMenu />
      </MemoryRouter>
    );
    expect(screen.getByRole("mobile-toggle")).toBeInTheDocument();
  });

  it("mobile nav menu should be hidden by default", function () {
    render(
      <MemoryRouter>
        <NavigationMenu />
      </MemoryRouter>
    );
    expect(screen.getByRole("mobile-nav-menu")).toHaveClass("hidden");
  });

  it("click mobile nav toggle button to display mobile nav menu", async function () {
    render(
      <MemoryRouter>
        <NavigationMenu />
      </MemoryRouter>
    );

    userEvent.click(screen.getByRole("mobile-toggle"));

    await waitFor(() => {
      expect(screen.getByRole("mobile-nav-menu")).toHaveClass("block");
    });
  });

  it("click mobile nav toggle button again hide mobile nav menu", async function () {
    render(
      <MemoryRouter>
        <NavigationMenu />
      </MemoryRouter>
    );

    userEvent.click(screen.getByRole("mobile-toggle"));
    userEvent.click(screen.getByRole("mobile-toggle"));

    await waitFor(() => {
      expect(screen.getByRole("mobile-nav-menu")).toHaveClass("hidden");
    });
  });

  it.each(["Home", "Posts", "Portfolio", "About me"])(
    "navigation bar should have desktop nav item '%s'",
    (name) => {
      render(
        <MemoryRouter>
          <NavigationMenu />
        </MemoryRouter>
      );
      expect(
        screen.getByText(name, {
          selector: ".desktop-nav-item",
        })
      ).toBeInTheDocument();
    }
  );

  it.each(["Home", "Posts", "Portfolio", "About me"])(
    "navigation bar should have mobile nav item '%s'",
    (name) => {
      render(
        <MemoryRouter>
          <NavigationMenu />
        </MemoryRouter>
      );
      expect(
        screen.getByText(name, {
          selector: ".mobile-nav-item",
        })
      ).toBeInTheDocument();
    }
  );
});
