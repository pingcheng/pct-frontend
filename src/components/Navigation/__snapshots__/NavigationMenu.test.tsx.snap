// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <NavigationMenu> should render correctly 1`] = `"<nav class=\\"bg-gray-700\\"><div class=\\"container-body\\"><div class=\\"relative flex items-center justify-between h-16\\"><div class=\\"absolute inset-y-0 left-0 flex items-center sm:hidden\\"><button class=\\"inline-flex items-center justify-center p-2 rounded-md text-gray-400 hover:text-white hover:bg-gray-700 focus:outline-none focus:ring-2 focus:ring-inset focus:ring-white\\" aria-expanded=\\"false\\" role=\\"mobile-toggle\\"><span class=\\"sr-only\\">Open main menu</span><svg class=\\"block h-6 w-6\\" xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" aria-hidden=\\"true\\"><path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M4 6h16M4 12h16M4 18h16\\"></path></svg><svg class=\\"hidden h-6 w-6\\" xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" aria-hidden=\\"true\\"><path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M6 18L18 6M6 6l12 12\\"></path></svg></button></div><div class=\\"flex-1 flex items-center justify-center sm:items-stretch sm:justify-start\\"><div class=\\"flex-shrink-0 flex items-center\\"><div class=\\"font-bold text-white\\" style=\\"border-top: 3px solid white; border-bottom: 3px solid white;\\">Ping Cheng</div></div><div class=\\"hidden sm:block sm:ml-6\\"><div class=\\"flex space-x-4\\"><a aria-current=\\"page\\" class=\\"nav-item desktop-nav-item active\\" href=\\"/\\">Home</a><a class=\\"nav-item desktop-nav-item\\" href=\\"/posts\\">Posts</a><a class=\\"nav-item desktop-nav-item\\" href=\\"/portfolio\\">Portfolio</a><a class=\\"nav-item desktop-nav-item\\" href=\\"/about\\">About me</a></div></div></div></div></div><div class=\\"sm:hidden hidden\\" role=\\"mobile-nav-menu\\"><div class=\\"px-2 pt-2 pb-3 space-y-1\\"><a aria-current=\\"page\\" class=\\"nav-item mobile-nav-item active\\" href=\\"/\\">Home</a><a class=\\"nav-item mobile-nav-item\\" href=\\"/posts\\">Posts</a><a class=\\"nav-item mobile-nav-item\\" href=\\"/portfolio\\">Portfolio</a><a class=\\"nav-item mobile-nav-item\\" href=\\"/about\\">About me</a></div></div></nav>"`;
